module/VuFind/src/VuFind/View/Helper/Root/Icon.php

```diff
@@ -0,0 +1,184 @@
+    protected $defaultSet;
+
+    /**
+     * Transforming map
+     *
+     * @var array
+     */
+    protected $iconMap;
+
+    /**
+     * Cache for icons
+     *
+     * @var StorageInterface
+     */
+    protected $cache;
+
+    /**
+     * Constructor
+     *
+     * @param ThemeInfo $themeInfo Theme info helper
+     */
+    public function __construct(ThemeInfo $themeInfo, CacheManager $cacheManager)
+    {
+        $this->config = $themeInfo->getMergedConfig('icons', true);
+        $this->defaultSet = $this->config['defaultSet'] ?? 'FontAwesome';
+        $this->iconMap = $this->config['aliases'] ?? [];
+        $this->cache = $cacheManager->getCache('object', 'iconHelper');
+    }
+
+    /**
+     * Map icon to set. Add prefix, return with set and template.
+     * Broken out for easier customization.
+     *
+     * @param string $name Which icon?
+     *
+     * @return array
+     */
+    protected function mapIcon(string $name): array
+    {
+        $icon = $this->iconMap[$name] ?? $name;
+        $set = $this->defaultSet;
+
+        // Override set from config (ie. FontAwesome:icon)
+        if (strpos($icon, ':') !== false) {
+            [$set, $icon] = explode(':', $icon, 2);
+        }
+
+        // Find set in theme.config.php
+        $setConfig = $this->config['sets'][$set] ?? [];
+        $template = $setConfig['template'] ?? $set;
+        $prefix = $setConfig['prefix'] ?? '';
+
+        return [$prefix . $icon, $set, $template];
+    }
+
+    /**
+     * Reduce extra parameters to one attribute string.
+     * Broken out for easier customization.
+     *
+     * @param array          $extra   Just extra HTML attributes for now
+     * @param EscapeHtmlAttr $escAttr EscapeHtmlAttr view helper
+     *
+     * @return string
+     */
+    protected function compileAttrs(array $extra, EscapeHtmlAttr $escAttr): string
+    {
+        $attrs = '';
+        foreach ($extra as $key => $val) {
+            $attrs .= ' ' . $key . '="' . $escAttr($val) . '"';
+        }
+        return $attrs;
+    }
+
+    /**
+     * Create a unique key for icon names and extra attributes
+     *
+     * @return string
+     */
+    protected function cacheKey(string $name, $extra = []): string
+    {
+        if (empty($extra)) {
+            return $name;
+        }
+        ksort($extra);
+        return $name . '+' . md5(json_encode($extra));
+    }
+
+    /**
+     * Returns inline HTML for icon
+     *
+     * @param string $name  Which icon?
+     * @param array  $extra Just extra HTML attributes for now
+     *
+     * @return string
+     */
+    public function __invoke(string $name, $extra = []): string
+    {
+        $cacheKey = $this->cacheKey($name, $extra);
+        $cached = $this->cache->getItem($cacheKey);
+
+        if ($cached == null) {
+            [$icon, $set, $template] = $this->mapIcon($name);
+
+            // Compile additional HTML attributes
+            $escAttr = $this->getView()->plugin('escapeHtmlAttr');
```

Comment:
Would it be better to assign $escAttr during construct so, it would not be required as a parameter to a function?

Reply:
I agree; since we have a custom factory already anyway, we might as well make this a hard dependency. That will make test-writing easier, too.

Reply:
Done!